File: tui/src/scala/tui/Layout.scala
package tui

import tui.cassowary.Strength.{REQUIRED, WEAK}
import tui.cassowary.WeightedRelation._
import tui.cassowary.operators._
import tui.cassowary.{Constraint => CassowaryConstraint, Expression, Solver, Variable}
import tui.internal.ranges

import scala.collection.mutable

/** @param direction
  * @param margin
  * @param constraints
  *   Whether the last chunk of the computed layout should be expanded to fill the available space.
  * @param expand_to_fill
  */
class Layout(
    direction: Direction = Direction.Vertical,
    margin: Margin = Margin.None,
    expandToFill: Boolean = true
)(widgets: (Constraint, Widget)*)
    extends Widget {

  override def render(area: Rect, buf: Buffer): Unit = {
    val constraints = widgets.map(_._1).toArray
    val chunks = Layout.cached(area, constraints, margin, direction, expandToFill)

    ranges.range(0, widgets.length) { i =>
      val (_, widget) = widgets(i)
      val chunk = chunks(i)
      widget.render(chunk, buf)
    }
  }
}

object Layout {
  def detailed(direction: Direction = Direction.Vertical, margin: Margin = Margin.None, expandToFill: Boolean = true)(widgets: (Constraint, Widget)*) =
    new Layout(direction, margin, expandToFill)(widgets: _*)

  def apply(direction: Direction = Direction.Vertical, margin: Margin = Margin.None, expandToFill: Boolean = true)(widgets: Widget*) =
    new Layout(direction, margin, expandToFill)(widgets.map(w => Constraint.Ratio(1, widgets.length) -> w): _*)

  private val LAYOUT_CACHE = mutable.HashMap.empty[(Rect, Array[Constraint], Margin, Direction, Boolean), Array[Rect]]

  /** A container used by the solver inside split
    */
  case class Element(
      x: Variable = Variable(),
      y: Variable = Variable(),
      width: Variable = Variable(),
      height: Variable = Variable()
  ) {
    def left: Variable = x

    def top: Variable = y

    def right: Expression = x + width

    def bottom: Expression = y + height
  }

  def cached(area: Rect, constraints: Array[Constraint], margin: Margin, direction: Direction, expandToFill: Boolean): Array[Rect] =
    Layout.LAYOUT_CACHE.getOrElseUpdate(
      (area, constraints, margin, direction, expandToFill),
      split(area, constraints, margin, direction, expandToFill)
    )

  def split(area: Rect, constraints: Array[Constraint], margin: Margin, direction: Direction, expandToFill: Boolean): Array[Rect] = {
    val solver = new Solver
    val vars: mutable.Map[Variable, (Int, Int)] = mutable.Map.empty
    val elements = constraints.map(_ => new Element)
    val results: Array[Rect] = constraints.map(_ => Rect.default)

    val dest_area = area.inner(margin)
    ranges.range(0, elements.length) { i =>
      val e = elements(i)
      vars.update(e.x, (i, 0))
      vars.update(e.y, (i, 1))
      vars.update(e.width, (i, 2))
      vars.update(e.height, (i, 3));
    }

    val ccs = Array.newBuilder[CassowaryConstraint]
    ccs.sizeHint(elements.length * 4 + constraints.length * 6)

    elements.foreach { elt =>
      ccs += (elt.width | GE(REQUIRED) | 0.0)
      ccs += (elt.height | GE(REQUIRED) | 0.0)
      ccs += (elt.left | GE(REQUIRED) | dest_area.left.toDouble)
      ccs += (elt.top | GE(REQUIRED) | dest_area.top.toDouble)
      ccs += (elt.right | LE(REQUIRED) | dest_area.right.toDouble)
      ccs += (elt.bottom | LE(REQUIRED) | dest_area.bottom.toDouble);
    }

    elements.headOption foreach { first =>
      val c = direction match {
        case Direction.Horizontal => first.left | EQ(REQUIRED) | dest_area.left.toDouble
        case Direction.Vertical   => first.top | EQ(REQUIRED) | dest_area.top.toDouble
      }
      ccs += c
    }

    if (expandToFill) {
      elements.lastOption.foreach { last =>
        val c = direction match {
          case Direction.Horizontal => last.right | EQ(REQUIRED) | dest_area.right.toDouble
          case Direction.Vertical   => last.bottom | EQ(REQUIRED) | dest_area.bottom.toDouble
        }
        ccs += c
      }
    }

    direction match {
      case Direction.Horizontal =>
        elements.sliding(2).foreach {
          case Array(one, two) => ccs += ((one.x + one.width) | EQ(REQUIRED) | two.x);
          case _               => // ignore if only one
        }

        ranges.range(0, constraints.length) { i =>
          ccs += (elements(i).y | EQ(REQUIRED) | dest_area.y.toDouble)
          ccs += (elements(i).height | EQ(REQUIRED) | dest_area.height.toDouble)
          val size = constraints(i)
          ccs += (size match {
            case Constraint.Length(v) =>
              elements(i).width | EQ(WEAK) | v.toDouble
            case Constraint.Percentage(v) =>
              elements(i).width | EQ(WEAK) | ((v * dest_area.width).toDouble / 100.0)
            case Constraint.Ratio(n, d) =>
              elements(i).width | EQ(WEAK) | (dest_area.width.toDouble * n.toDouble / d.toDouble)
            case Constraint.Min(v) =>
              elements(i).width | GE(WEAK) | v.toDouble
            case Constraint.Max(v) =>
              elements(i).width | LE(WEAK) | v.toDouble
          })
        }
      case Direction.Vertical =>
        elements.sliding(2).foreach {
          case Array(one, two) => ccs += ((one.y + one.height) | EQ(REQUIRED) | two.y);
          case _               => // ignore if only one
        }
        ranges.range(0, constraints.length) { i =>
          ccs += (elements(i).x | EQ(REQUIRED) | dest_area.x.toDouble)
          ccs += (elements(i).width | EQ(REQUIRED) | dest_area.width.toDouble)
          val size = constraints(i)
          ccs += (size match {
            case Constraint.Length(v) =>
              elements(i).height | EQ(WEAK) | v.toDouble
            case Constraint.Percentage(v) =>
              elements(i).height | EQ(WEAK) | ((v * dest_area.height).toDouble / 100.0)
            case Constraint.Ratio(n, d) =>
              elements(i).height | EQ(WEAK) | (dest_area.height.toDouble * n.toDouble / d.toDouble)
            case Constraint.Min(v) =>
              elements(i).height | GE(WEAK) | v.toDouble
            case Constraint.Max(v) =>
              elements(i).height | LE(WEAK) | v.toDouble
          });
        }

    }

    solver.add_constraints(ccs.result()) match {
      case Left(err) => sys.error(s"Error while adding constraints: $err")
      case Right(()) => ()
    }

    solver.fetch_changes().foreach { case (v, value0) =>
      val (index, attr) = vars(v)
      val value = math.max(0, value0.toInt)
      val rect = results(index)
      attr match {
        case 0 => results(index) = rect.copy(x = value)
        case 1 => results(index) = rect.copy(y = value)
        case 2 => results(index) = rect.copy(width = value)
        case 3 => results(index) = rect.copy(height = value)
        case _ => ()
      }
    }

    if (expandToFill) {
      // Fix imprecision by extending the last item a bit if necessary
      if (results.nonEmpty) {
        val lastIdx = results.length - 1
        val last = results(lastIdx)
        val updated = direction match {
          case Direction.Horizontal => last.copy(height = dest_area.bottom - last.y)
          case Direction.Vertical   => last.copy(width = dest_area.right - last.x)
        }
        results(lastIdx) = updated
      }
    }
    results
  }
}

File: tui/src/java/tui/Layout.java
package tui;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import tui.cassowary.CassowaryOps;
import tui.cassowary.Constraint;
import tui.cassowary.Either;
import tui.cassowary.Expression;
import tui.cassowary.Solver;
import tui.cassowary.Strength;
import tui.cassowary.Variable;
import tui.cassowary.VariableChange;
import tui.cassowary.WeightedRelation;
import tui.internal.Ranges;

public final class Layout {
  public final Direction direction;
  public final Margin margin;
  public final tui.Constraint[] constraints;
  public final boolean expandToFill;

  public Layout(
      Direction direction,
      Margin margin,
      tui.Constraint[] constraints,
      boolean expandToFill) {
    this.direction = direction;
    this.margin = margin;
    this.constraints = constraints;
    this.expandToFill = expandToFill;
  }

  public Rect[] split(Rect area) {
    CacheKey key = new CacheKey(area, this);
    Rect[] cached = LAYOUT_CACHE.get(key);
    if (cached != null) return cached;
    Rect[] computed = Layout.split(area, this);
    LAYOUT_CACHE.put(key, computed);
    return computed;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) return true;
    if (!(obj instanceof Layout other)) return false;
    if (direction != other.direction) return false;
    if (!margin.equals(other.margin)) return false;
    if (expandToFill != other.expandToFill) return false;
    if (constraints.length != other.constraints.length) return false;
    for (int i = 0; i < constraints.length; i++) {
      if (!constraints[i].equals(other.constraints[i])) return false;
    }
    return true;
  }

  @Override
  public int hashCode() {
    int h = direction.hashCode();
    h = 31 * h + margin.hashCode();
    h = 31 * h + Boolean.hashCode(expandToFill);
    for (tui.Constraint c : constraints) {
      h = 31 * h + c.hashCode();
    }
    return h;
  }

  public record CacheKey(Rect area, Layout layout) {}

  public static final Map<CacheKey, Rect[]> LAYOUT_CACHE = new HashMap<>();

  /// A container used by the solver inside split
  public static final class Element {
    public final Variable x;
    public final Variable y;
    public final Variable width;
    public final Variable height;

    public Element() {
      this.x = Variable.create();
      this.y = Variable.create();
      this.width = Variable.create();
      this.height = Variable.create();
    }

    public Variable left() {
      return x;
    }

    public Variable top() {
      return y;
    }

    public Expression right() {
      return CassowaryOps.add(x, width);
    }

    public Expression bottom() {
      return CassowaryOps.add(y, height);
    }
  }

  public static Rect[] split(Rect area, Layout layout) {
    Solver solver = new Solver();
    Map<Variable, IntPair> vars = new HashMap<>();
    Element[] elements = new Element[layout.constraints.length];
    for (int i = 0; i < elements.length; i++) elements[i] = new Element();
    Rect[] results = new Rect[layout.constraints.length];
    for (int i = 0; i < results.length; i++) results[i] = Rect.DEFAULT;

    Rect destArea = area.inner(layout.margin);
    Ranges.range(
        0,
        elements.length,
        i -> {
          Element e = elements[i];
          vars.put(e.x, new IntPair(i, 0));
          vars.put(e.y, new IntPair(i, 1));
          vars.put(e.width, new IntPair(i, 2));
          vars.put(e.height, new IntPair(i, 3));
        });

    java.util.ArrayList<Constraint> ccs = new java.util.ArrayList<>();
    ccs.ensureCapacity(elements.length * 4 + layout.constraints.length * 6);

    for (Element elt : elements) {
      ccs.add(CassowaryOps.constraint(elt.width, WeightedRelation.GE(Strength.REQUIRED), 0.0));
      ccs.add(CassowaryOps.constraint(elt.height, WeightedRelation.GE(Strength.REQUIRED), 0.0));
      ccs.add(
          CassowaryOps.constraint(
              elt.left(), WeightedRelation.GE(Strength.REQUIRED), (double) destArea.left()));
      ccs.add(
          CassowaryOps.constraint(
              elt.top(), WeightedRelation.GE(Strength.REQUIRED), (double) destArea.top()));
      ccs.add(
          CassowaryOps.constraint(
              elt.right(), WeightedRelation.LE(Strength.REQUIRED), (double) destArea.right()));
      ccs.add(
          CassowaryOps.constraint(
              elt.bottom(), WeightedRelation.LE(Strength.REQUIRED), (double) destArea.bottom()));
    }

    if (elements.length > 0) {
      Element first = elements[0];
      Constraint c =
          switch (layout.direction) {
            case Horizontal ->
                CassowaryOps.constraint(
                    first.left(),
                    WeightedRelation.EQ(Strength.REQUIRED),
                    (double) destArea.left());
            case Vertical ->
                CassowaryOps.constraint(
                    first.top(),
                    WeightedRelation.EQ(Strength.REQUIRED),
                    (double) destArea.top());
          };
      ccs.add(c);
    }

    if (layout.expandToFill && elements.length > 0) {
      Element last = elements[elements.length - 1];
      Constraint c =
          switch (layout.direction) {
            case Horizontal ->
                CassowaryOps.constraint(
                    last.right(),
                    WeightedRelation.EQ(Strength.REQUIRED),
                    (double) destArea.right());
            case Vertical ->
                CassowaryOps.constraint(
                    last.bottom(),
                    WeightedRelation.EQ(Strength.REQUIRED),
                    (double) destArea.bottom());
          };
      ccs.add(c);
    }

    switch (layout.direction) {
      case Horizontal -> {
        for (int j = 0; j + 1 < elements.length; j++) {
          Element one = elements[j];
          Element two = elements[j + 1];
          ccs.add(
              CassowaryOps.constraint(
                  CassowaryOps.add(one.x, one.width),
                  WeightedRelation.EQ(Strength.REQUIRED),
                  two.x));
        }

        Ranges.range(
            0,
            layout.constraints.length,
            i -> {
              ccs.add(
                  CassowaryOps.constraint(
                      elements[i].y,
                      WeightedRelation.EQ(Strength.REQUIRED),
                      (double) destArea.y()));
              ccs.add(
                  CassowaryOps.constraint(
                      elements[i].height,
                      WeightedRelation.EQ(Strength.REQUIRED),
                      (double) destArea.height()));
              tui.Constraint size = layout.constraints[i];
              Constraint built =
                  switch (size) {
                    case tui.Constraint.Length l ->
                        CassowaryOps.constraint(
                            elements[i].width,
                            WeightedRelation.EQ(Strength.MEDIUM),
                            (double) l.l());
                    case tui.Constraint.Percentage p ->
                        CassowaryOps.constraint(
                            elements[i].width,
                            WeightedRelation.EQ(Strength.MEDIUM),
                            ((double) (p.p() * destArea.width())) / 100.0);
                    case tui.Constraint.Ratio r ->
                        CassowaryOps.constraint(
                            elements[i].width,
                            WeightedRelation.EQ(Strength.MEDIUM),
                            (double) destArea.width() * (double) r.num() / (double) r.den());
                    case tui.Constraint.Min m ->
                        CassowaryOps.constraint(
                            elements[i].width, WeightedRelation.GE(Strength.MEDIUM), (double) m.m());
                    case tui.Constraint.Max m ->
                        CassowaryOps.constraint(
                            elements[i].width, WeightedRelation.LE(Strength.MEDIUM), (double) m.m());
                  };
              ccs.add(built);
              switch (size) {
                case tui.Constraint.Min m ->
                    ccs.add(
                        CassowaryOps.constraint(
                            elements[i].width,
                            WeightedRelation.EQ(Strength.WEAK),
                            (double) m.m()));
                case tui.Constraint.Max m ->
                    ccs.add(
                        CassowaryOps.constraint(
                            elements[i].width,
                            WeightedRelation.EQ(Strength.WEAK),
                            (double) m.m()));
                default -> {}
              }
            });
      }
      case Vertical -> {
        for (int j = 0; j + 1 < elements.length; j++) {
          Element one = elements[j];
          Element two = elements[j + 1];
          ccs.add(
              CassowaryOps.constraint(
                  CassowaryOps.add(one.y, one.height),
                  WeightedRelation.EQ(Strength.REQUIRED),
                  two.y));
        }
        Ranges.range(
            0,
            layout.constraints.length,
            i -> {
              ccs.add(
                  CassowaryOps.constraint(
                      elements[i].x,
                      WeightedRelation.EQ(Strength.REQUIRED),
                      (double) destArea.x()));
              ccs.add(
                  CassowaryOps.constraint(
                      elements[i].width,
                      WeightedRelation.EQ(Strength.REQUIRED),
                      (double) destArea.width()));
              tui.Constraint size = layout.constraints[i];
              Constraint built =
                  switch (size) {
                    case tui.Constraint.Length l ->
                        CassowaryOps.constraint(
                            elements[i].height,
                            WeightedRelation.EQ(Strength.MEDIUM),
                            (double) l.l());
                    case tui.Constraint.Percentage p ->
                        CassowaryOps.constraint(
                            elements[i].height,
                            WeightedRelation.EQ(Strength.MEDIUM),
                            ((double) (p.p() * destArea.height())) / 100.0);
                    case tui.Constraint.Ratio r ->
                        CassowaryOps.constraint(
                            elements[i].height,
                            WeightedRelation.EQ(Strength.MEDIUM),
                            (double) destArea.height() * (double) r.num() / (double) r.den());
                    case tui.Constraint.Min m ->
                        CassowaryOps.constraint(
                            elements[i].height,
                            WeightedRelation.GE(Strength.MEDIUM),
                            (double) m.m());
                    case tui.Constraint.Max m ->
                        CassowaryOps.constraint(
                            elements[i].height,
                            WeightedRelation.LE(Strength.MEDIUM),
                            (double) m.m());
                  };
              ccs.add(built);
              switch (size) {
                case tui.Constraint.Min m ->
                    ccs.add(
                        CassowaryOps.constraint(
                            elements[i].height,
                            WeightedRelation.EQ(Strength.WEAK),
                            (double) m.m()));
                case tui.Constraint.Max m ->
                    ccs.add(
                        CassowaryOps.constraint(
                            elements[i].height,
                            WeightedRelation.EQ(Strength.WEAK),
                            (double) m.m()));
                default -> {}
              }
            });
      }
    }

    Either<tui.cassowary.AddConstraintError, Void> result = solver.addConstraints(ccs);
    if (result instanceof Either.Left<tui.cassowary.AddConstraintError, Void> l) {
      throw new RuntimeException("Error while adding constraints: " + l.value());
    }

    for (VariableChange vc : solver.fetchChanges()) {
      Variable v = vc.variable();
      double value0 = vc.value();
      IntPair pair = vars.get(v);
      int index = pair.first();
      int attr = pair.second();
      int value = Math.max(0, (int) value0);
      Rect rect = results[index];
      switch (attr) {
        case 0 -> results[index] = new Rect(value, rect.y(), rect.width(), rect.height());
        case 1 -> results[index] = new Rect(rect.x(), value, rect.width(), rect.height());
        case 2 -> results[index] = new Rect(rect.x(), rect.y(), value, rect.height());
        case 3 -> results[index] = new Rect(rect.x(), rect.y(), rect.width(), value);
        default -> {}
      }
    }

    if (layout.expandToFill) {
      // Fix imprecision by extending the last item a bit if necessary
      if (results.length > 0) {
        int lastIdx = results.length - 1;
        Rect last = results[lastIdx];
        Rect updated;
        switch (layout.direction) {
          case Horizontal ->
              updated = new Rect(last.x(), last.y(), last.width(), destArea.bottom() - last.y());
          case Vertical ->
              updated = new Rect(last.x(), last.y(), destArea.right() - last.x(), last.height());
          default -> updated = last;
        }
        results[lastIdx] = updated;
      }
    }
    return results;
  }

  private record IntPair(int first, int second) {}
}
